### ISSUE TITLE
Preview generation takes an unreasonably long time

### ISSUE THREAD

**Issue by NeversSync** [prismicio/prismic-gatsby#500]
### Versions

- gatsby-source-prismic: 5.1.0
- gatsby-plugin-prismic-previews: 5.0.3
- gatsby: 4.7.1

### Reproduction

The best I can provide is a description of our experience as our preview links are not shareable here. 

Previews generally take between 25-60 seconds to load causing people using them to assume they don't work or to stop using them because they are so slow. 

<summary>Most of our team has stopped using the preview function ever since we moved to the most recent major version (5 of both). This is a huge thorn in our sides and hugely reduces the value of the preview function of Prismic and Prismic in general. 

From what I understand, the preview action currently builds the requested document, processes any images on the page, and perhaps also tries to build pages that are linked from that document. If I'm mistaken, I'd love to see the detailed description of what is happening in the documentation. 

Are there any changes that can be implemented to meaningfully reduce the time it takes to build a preview? Are other users experiencing this as well? 
Some ideas: 

1. A way to opt out of building documents/pages other than the individual document requested. 
2. A way to opt in to having lower resolution on images for previews.
Without understanding the mechanisms under the hood, it's hard for me to think of other suggestions or to feel confident in the above. </summary>

<br>

Package.json:
![Screen Shot 2022-02-25 at 10 26 34 AM](https://user-images.githubusercontent.com/20482636/155768105-6b70dfae-5d45-4d4a-9d3b-8730d6f30b0b.png)

### Steps to reproduce
On any site with images and more than 150 pages, load a preview from the Prismic UI. 

### What is expected?
Previews should load in a reasonable time (less than 15 seconds?) and feel like a usable feature for all team members. 

### What is actually happening?
Previews generally take between 25-60 seconds to load causing people using them to assume they don't work or to stop using them because they are so slow.

**Comment by angeloashmore** [comment]
Hey @NeversSync, thanks for the report. Previews should not take that long to resolve. Generally speaking, previews should be ready within a couple of seconds.

Do you mind sharing a few details about your site and Prismic repository?

- **How many documents do you have in your repository?**
   The preview plugin requests all of your documents at the start of a preview session. By default, each requests includes 100 documents. For a repository of ~150 documents, you should only see three `/api/v2` API requests.

- **Is the long load time on the preview resolver page (usually `/preview`), or is it on the actual page being previewed?**
   I assume it's on the document's page, but it's worth ruling out `withPrismicPreviewResolver()`.

- **In your network tab in your browser's DevTools, do you see any duplicate or looping requests?**
   I'm wondering if maybe there's a bug somewhere causing the plugin to restart its process.

In case you would like to see an overview of how the plugin system works, there's a high-level overview in the ARCHITECTURE.md document: https://github.com/prismicio/prismic-gatsby/blob/main/ARCHITECTURE.md#during-a-preview-session

I hope we can figure out the issue. Previews are an essential feature and it definitely should not take 25-60 seconds. Thanks for the help in figuring this out. 🙂

**Comment by kb1995** [comment]
Just wanted to share that I also started experiencing this problem for one of my clients. It wasn't a problem until a month or two ago, so probably something related to a recent upgrade. 

- How many documents do you have in your repository? -> around ~150 documents
- Is the long load time on the preview resolver page (usually /preview), or is it on the actual page being previewed? -> the actual page, so I'd see the modal being loaded
- In your network tab in your browser's DevTools, do you see any duplicate or looping requests? -> I haven't checked yet

**Comment by angeloashmore** [comment]
Thanks for the additional info @kb1995. I'll take a closer look to figure out what's going on.

**Comment by NeversSync** [comment]
@angeloashmore Thanks for the quick reply. It's helpful to know that our wait time is abnormal, and also to have the link to the documentation on the workings of the preview plugin system. 

Answers to your questions: 

**- How many documents do you have in your repository?** -> around 715 documents

**- Is the long load time on the preview resolver page (usually /preview), or is it on the actual page being previewed?** -> It is on the actual page with the 'fetch preview' modal visible

**- In your network tab in your browser's DevTools, do you see any duplicate or looping requests?** -> I don't see any duplicate, but I do see a lot of requests coming in that are taking > 500ms. Here is a screenshot of the network tab sorted by time. I also downloaded an HAR recording but it appears I can't attach that here. LMK if that would be helpful. I really appreciate your time. 

![Screen Shot 2022-02-28 at 11 12 10 AM](https://user-images.githubusercontent.com/20482636/156044411-be7eef1d-2612-47b4-b476-56fb30599a5c.png)

**Comment by angeloashmore** [comment]
It appears to happen due to the API, not necessarily from `gatsby-plugin-prismic-previews`. The load times are exacerbated by the preview plugin's behavior of fetching all documents on entry (this is unlikely to change and has generally worked well in the past).

I've raised the issue with Prismic's API team to understand if there were recent architecture changes that could be causing this. I'll share more info here once I learn more. Thanks for your help @NeversSync @kb1995!

**Comment by angeloashmore** [comment]
Hey @NeversSync @kb1995, the API team said they found an issue and fixed it (sorry, I don't have details on what the issue was).

Could you test a preview again and let me know if the load times are the same or improved? Thanks

**Comment by kb1995** [comment]
@angeloashmore Thanks for the quick response. Unfortunately, I just tried it and the preview takes the same amount of time as before

**Comment by NeversSync** [comment]
@angeloashmore I just tried a preview again on a pretty small document, and it took ~30 seconds, about the same as it was before.

**Comment by angeloashmore** [comment]
Thanks for testing. I shared your responses to the team. Once I hear back, I'll give an update here. 🙂

**Comment by Baztoune** [comment]
Hi guys, I already commented on another issue, but my colleague found out about this dedicated one, so sorry about the duplicate. We are experiencing a very similar problem, with up to date libs, and ~700 prismic documents.
The network tab reveals 8 prismic API requests as expected, which take from 1 to 2sec each. That doesn't add up to the ~60 seconds with frozen UI on my Apple M1 (worse on other CPUs). 
So it looks like it comes from parsing or some CPU intensive function. I'm not a profiling expert, but the first profiling indicates `normalizeDocument` as a culprit. Do you think this is the right direction ?
<img width="1440" alt="Screenshot 2022-03-04 at 16 51 19" src="https://user-images.githubusercontent.com/288581/156795622-5b148aed-ab83-4639-83d1-0f15f791149c.png">

**Comment by Baztoune** [comment]
Could it be the `serializePath` function added in #466 ? 
https://github.com/prismicio/prismic-gatsby/blob/800414e4027775d885493e715d0bc6005bfddbda/packages/gatsby-source-prismic/src/lib/getTypePath.ts#L24
I also noticed that with `yarn start` (which uses `NODE_ENV=development`) the preview loading takes ~10sec (network+processing which is around the time before the upgrade) whereas with `yarn build && yarn serve` (which uses `NODE_ENV=production`) the preview is ~60sec, which makes me think it could be the `md5` function.
https://github.com/prismicio/prismic-gatsby/blob/800414e4027775d885493e715d0bc6005bfddbda/packages/gatsby-source-prismic/src/runtime/serializePath.ts#L4
Do you notice the same behaviour ?

**Comment by NeversSync** [comment]
@Baztoune thank you for digging in to this and providing these super helpful insights. 

To answer your question, I just tested this on a mid-sized document with a few images and can confirm that I'm seeing the same behavior as you are:

- `gatsby develop` preview took 22 seconds 
- `gatsby build && gatsby serve` preview took 40 seconds

Curious to hear thoughts from @angeloashmore and the Prismic team.

**Comment by Baztoune** [comment]
Thank you for your feedback ! From my understanding, the processing time actually depends on the whole amount of documents you have in prismic, not just the one you're trying to preview.  The plugin "normalizes" all these documents (including a md5 hash of its path when `NODE_ENV=production`), and it looks like it normalizes all these documents subtree too (see `normalizeDocumentSubtree` recursive function)
https://github.com/prismicio/prismic-gatsby/blob/800414e4027775d885493e715d0bc6005bfddbda/packages/gatsby-source-prismic/src/lib/normalizeDocumentSubtree.ts#L172-L181

So my guess is that each document is normalized multiple times, because it's part of other documents subtrees.

I didn't have time to try yet and I have no knowledge with fp-ts, but I wanted to try to add a simple cache / localstorage / memoize (using https://github.com/ds300/patch-package), at least for the `serializePath` function. Actually, maybe the whole `normalizeDocument` result could be cached ? That would really speed things up A LOT.

**Comment by samlogan** [comment]
+1

**Comment by angeloashmore** [comment]
Hey everyone, thank you for the detailed triaging and reports.

It looks like you are all correct; the slowness is not only a result of waiting on network requests. It is also due to excessive data processing.

From a very high level, this is what the preview plugin is doing:

1. **Fetches all documents from the Prismic repository.**
   This is necessary because your GraphQL queries could reference any document, even ones that are not directly linked to the previewed document. When you consider that Prismic Releases (multiple document changes) can be previewed, those extra documents also need to be fetched.

2. **"Normalizes" each document to be shaped like Gatsby's GraphQL API
   This is necessary since data fetched at runtime from Prismic's API does not appear the same as it does from Gatsby. A Rich Text field, for example, needs to process its content and make `text`, `html`, and `richText` properties available, just like in the Gatsby GraphQL API.

   Each document is only processed once synchronously.

   In the case of linked documents, they are made available using Proxies. This is done do prevent eagerly evaluated infinitely nested properties.

3. **Replaces any document in the `data` prop containing a matching `_previewable` property with a normalized object**
    This is how the previewed document data gets injected into the component.

There are some strategies we can take to reduce Step 2 (document normalization), such as making all normalization lazy via Proxies.

I'm not sure how we can improve on Step 1 (fetch all documents). In a previous implementation of the preview plugin, only the previewed document and its linked documents were fetched. This could potentially lead to a waterfall of requests, each relying on the one before it. This also wouldn't support Prismic Releases since there is no way currently to fetch a list of documents in a Release.

Until a larger rework of this flow is possible, there may be quick wins we can make to speed up the process. One example is removing MD5 processing, like @Baztoune suggests. I'll need to do more research on exactly what is requiring so much processing.

Thank you again everyone for the details, and I 100% understand that previewing with `gatsby-plugin-prismic-previews` is not practically usable for larger projects at this time.

**Comment by Baztoune** [comment]
Thank you for your detailed answer, this helps to understand what happens behind the scene. FYI we kind of solved the performance issue by forking `tiny-hashes` lib in order to override the `md5` function (it now returns the raw input string, and we're back from 70 seconds processing to 10 seconds). Then yarn allows to link the github fork, which is easy to work with, and will be easy to remove when a permanent fix is available.
This is a dirty fix, but it was really hard to patch `gatsby-source-prismic` in our monorepo, and we haven't noticed any issue for now.
Can you explain why paths are hashed with md5 in production mode ? As a workaround, could it be disabled with an option instead of relying on `NODE_ENV` ?

**Comment by AdrianGuery** [comment]
Any update on this issue ? 

On a Gatsby prismic project with more than 450 documents, a preview takes more than 3 minutes.

@Baztoune is it possible to have access to the source code of your patch?

**Comment by Baztoune** [comment]
@AdrianGuery our "patch" is just a fork of https://github.com/jbt/tiny-hashes with `md5.js` file content replaced by
```
export default function md5(s) {  return s; }
```
We use it by overriding yarn resolution and pointing to our fork. Depending on your package manager, it could be different.
```
"resolutions": {
    "tiny-hashes": "git+https://git@github.com/xxxxxxx/tiny-hashes.git"
}
```
Not the cleanest workaround but it works.

**Comment by AdrianGuery** [comment]
Thanks a lot @Baztoune !

**Comment by AdrianGuery** [comment]
Unfortunately, it improves the preview generation only in local. 
So I tried to limit the fetching of documents, and it works for us. In fact, we do not used the release system in Primsic. It takes now 10sec instead of 180+sec

There is the modification : https://github.com/AdrianGuery/prismic-gatsby/commit/72b0ba7f3d094df339583d5077c31930fb9a6f05

**Comment by mrseanbaines** [comment]
Also experiencing this. Mentioned a few months ago [here](https://github.com/prismicio/prismic-gatsby/issues/477#issuecomment-1025553968) but had no response

**Comment by kilinkis** [comment]
We are also having the same issue

**Comment by samlogan** [comment]
Be great to come up with a solution for this 🙏 We have a few big sites using Prismic and they're unable to preview content

**Comment by alextrepa** [comment]
+1. We have the same issue. Previews were taking multiple minutes. At least, we were able to implement @Baztoune workaround (thank you btw) and get them back to 20-30 seconds.

**Comment by angeloashmore** [comment]
Hey everyone, sorry for the delay in addressing this issue. I will be working on this and will publish a fix soon. I’ll post here once there is a new version to test.

Thank you!

**Comment by angeloashmore** [comment]
Hey everyone, could you try out this alpha version and let me know if it resolves the performance issue? It implements @Baztoune's `md5`-related fix across the whole plugin.

- `gatsby-source-prismic`: `v5.2.7-alpha.0`

`gatsby-plugin-prismic-preview` should use the updated version automatically once the updated source plugin is installed.

For more details, see #506.

**Comment by matijamrkaic** [comment]
Thank you @angeloashmore! Confirming that the previewing is much faster now!

**Comment by NeversSync** [comment]
Working much better now @angeloashmore! Thank you!

**Comment by angeloashmore** [comment]
Thanks for testing! 🙂

The fix has been merged and publish. Please update to the latest version of the Gatsby plugins when you are able to.

- `gatsby-source-prismic`: `v5.2.7`
- `gatsby-plugin-prismic-previews`: `v5.2.7`

I'm going to close this issue as the MD5 fix brings previews back into an "expected" performance level. I will be looking at different strategies that will fundamentally improve preview performance. There currently isn't a timeline for this effort, but it is being worked on.

Thanks again to @Baztoune and @NeversSync for the triaging and solutions. 🙌

**Comment by NeversSync** [comment]
Thank you @angeloashmore !!!

**Comment by kilinkis** [comment]
hi @angeloashmore, don't wanna ruin the party here, but we are still experiencing very long times after upgrading, even more than before.

Asking your questions from above:

*  How many documents do you have in your repository?
   About 6000 (about 300 per locale)
*Is the long load time on the preview resolver page (usually /preview), or is it on the actual page being previewed?
   This happens on the URL of the previewed document
* In your network tab in your browser's DevTools, do you see any duplicate or looping requests?
   On the network tab, I see that there are requests for all the documents of the current locale, and once that is done, nothing else happens in the Network tab, and the Loading Preview screen remains. Only after like 2 more minutes, the modal window clears and I can see my page with the new content coming from the preview.

Our versions are:
"gatsby-source-prismic": "5.2.7",
"gatsby-plugin-prismic-previews": "5.2.8",
"gatsby": "4.15.1",

Do you have any hints? thanks!
